File: src/main/java/com/vc/hard/L472.scala
package com.vc.hard

object WordOrderingByLength extends Ordering[String] {
  def compare(x: String, y: String): Int = x.length.compareTo(y.length)
}
object L472 {
  def findAllConcatenatedWordsInADict(words: Array[String]): List[String] = {
    import scala.collection.mutable
    import scala.util.Sorting.quickSort

    //Sort by length in ASC order
    quickSort(words)(WordOrderingByLength)

    def canForm(word: String, set: mutable.HashSet[String]): Boolean = {
      if(set.isEmpty) return false
      val n = word.length
      val dp = new Array[Boolean](n + 1)
      dp(0) = true
      (1 to n).foreach(i => {
        import scala.util.control.Breaks._
        breakable {
          (0 until i).foreach(j => {
            if(dp(j) && set.contains(word.substring(j, i))) {
              dp(i) = true
              break
            }
          })
        }
      })
      dp(n)
    }

    val list = new mutable.ListBuffer[String]()
    val set = new mutable.HashSet[String]()
    words.foreach(word => {
      if(canForm(word, set)) list += word
      set += word
    })
    list.toList
  }
}

File: src/main/java/com/vc/hard/L472.java
package com.vc.hard;

import java.util.*;

class L472 {

    static class WordComparator implements Comparator<String> {
        public int compare(String s1, String s2) {
            return s1.length() - s2.length();
        }
    }

    public List<String> findAllConcatenatedWordsInADict(String[] words) {
        Arrays.sort(words, new WordComparator());
        HashSet<String> set = new HashSet<>();
        List<String> res = new ArrayList<>();
        for(String word: words) {
            if(canForm(word, set)) res.add(word);
            set.add(word);
        }
        return res;
    }

    private boolean canForm(String word, HashSet<String> set) {
        if(set.isEmpty()) return false;
        int n = word.length();
        boolean[] dp = new boolean[n + 1];
        dp[0] = true;
        for(int i = 0; i <= n; i++) {
            for(int j = 0; j < i; j++) {
                if(dp[j] && set.contains(word.substring(j, i))) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[n];
    }
}
